### ISSUE TITLE
Mobile IAP causes crashes on Android

### ISSUE THREAD

**Issue by KekGames** [Scirra/Construct-3-bugs#3513]
## Problem description
 IAP plugin causes crashes on Android (crash report included)

**Affected browsers/platforms:** Android

## System details

<details><summary>View details</summary>

java.lang.NullPointerException: 
  at cc.fovea.PurchasePlugin.lambda$queryPurchases$1 (PurchasePlugin.java:300)
  at cc.fovea.-$$Lambda$PurchasePlugin$ERhMchr2fWN0IfG8YEXLUlz_LzM.run (Unknown Source:2)
  at cc.fovea.PurchasePlugin.executeServiceRequest (PurchasePlugin.java:853)
  at cc.fovea.PurchasePlugin.queryPurchases (PurchasePlugin.java:286)
  at cc.fovea.PurchasePlugin.onStart (PurchasePlugin.java:685)
  at org.apache.cordova.PluginManager.onStart (PluginManager.java:275)
  at org.apache.cordova.CordovaWebViewImpl.handleStart (CordovaWebViewImpl.java:466)
  at org.apache.cordova.CordovaActivity.onStart (CordovaActivity.java:303)
  at android.app.Instrumentation.callActivityOnStart (Instrumentation.java:1433)
  at android.app.Activity.performStart (Activity.java:7924)
  at android.app.ActivityThread.handleStartActivity (ActivityThread.java:3328)
  at android.app.servertransaction.TransactionExecutor.performLifecycleSequence (TransactionExecutor.java:221)
  at android.app.servertransaction.TransactionExecutor.cycleToPath (TransactionExecutor.java:201)
  at android.app.servertransaction.TransactionExecutor.executeLifecycleState (TransactionExecutor.java:173)
  at android.app.servertransaction.TransactionExecutor.execute (TransactionExecutor.java:97)
  at android.app.ActivityThread$H.handleMessage (ActivityThread.java:2040)
  at android.os.Handler.dispatchMessage (Handler.java:107)
  at android.os.Looper.loop (Looper.java:221)
  at android.app.ActivityThread.main (ActivityThread.java:7520)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:539)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:950)

</details>

**Comment by AshleyScirra** [comment]
@shortercode - we don't use `cc.fovea.PurchasePlugin` any more do we? This may be a third-party addon (in which case we can't offer support).

@KekGames - please note that in general we are unable to help unless you follow all the bug report guidelines. Often it is impossible to figure out the cause from just a crash report.

**Comment by shortercode** [comment]
This is the official plugin, we're forked off cordova-plugin-purchase which is maintained by fovea. It looks like this issue has been found and fixed upstream although they haven't published it yet.

https://github.com/j3k0/cordova-plugin-purchase/issues/946

I'll take a look at merging in pending fixes.

**Comment by KekGames** [comment]
@shortercode  so can it be fixed, or does it completely depend on fovea? It's very important to avoid crashes for better listing and featuring opportunities on Google Play.

**Comment by shortercode** [comment]
We can merge the fix over, but it's likely to have some other changes as well so I will need to review the changes.

**Comment by Boulerzzz** [comment]
@shortercode How does this affect users when they try to purchase? I tried myself and had had no problem but my developer console is filled with reports of these crashes? I get tons every day.

**Comment by KekGames** [comment]
@shortercode can we expect the fix any time soon?

**Comment by shortercode** [comment]
@KekGames the merge was delayed because of the holidays, sorry about that. I've published a new revision ( `construct-mobile-purchase@1.2.1` ) which includes the patches. Builds via the build service will use the new version automatically. If you are building from a cordova project you will likely have to update the version number of the plugin in the config.xml file before you see the changes.

I cannot guarantee that this resolves the issues, I had not been able replicate them myself but this does include a patch from upstream for what appears to be the same issue. 

Closing this as assumed fixed.
